### PR TITLE
Enable GCToOSInterface to be defined on the GC side of the GC interface

### PR DESCRIPTION
**Pull request by swgillespie**
This PR presents the framework and a limited implementation of `GCToOSInterface` being implemented by direct calls to platform libraries instead of travelling through the CoreCLR PAL.

The primary motivation for this change is that, if we want to use the GC in a standalone manner, all channels of communication between the GC and the CLR must be through versioned interfaces. For things like `GCToEEInterface`, this makes sense, as the GC is interacting with the CLR itself. However, this makes less sense for `GCToOSInterface`, since the CLR implementation of `GCToOSInterface` forwards its implementation on to the CoreCLR PAL without doing much processing of its own. Furthermore, since each interface must be versioned, which is a major compatibility hazard, especially if we want to do new and interesting things with the GC vis-a-vis the OS, through huge pages or `madvise` or the like. Since the OS surface area covered by `GCToOSInterface` is limited, it is reasonable to implement this interface in the GC itself so that the CLR does not need to worry about how the GC talks to the OS.

The changes in this PR provide the foundation for this idea by providing an implementation of `GCToOSInterface` that compiles alongside the GC. It's important to note that this PR is *not* feature-complete with regards to the existing interface. The interface implementation presented in this PR is the minimum to pass tests on Windows and compile on Unixes. In particular, the following things are not implemented  on Windows (but will be in future PRs soon):

1. Processor detection and other `cpuid`-based tasks (see e.g. `GCToOSInterface::GetLargestOnDieCacheSize`, `GCToOSInterface::GetLogicalCpuCount`)
2. Windows version detection (see e.g. `RunningOnWindows8`)
3. Processor affinity (see e.g. `GCToOSInterface::GetCurrentProcessCpuCount`)

This implementation came mostly from CoreRT's implementation of `GCToOSInterface` with some additional code from CoreCLR's implementation today.

An important thing to note is that each OS's implementation of `GCToOSInterface` has to include system headers in order to function. This is not a problem on Windows, but Unixes pass `-nostdinc` to Clang when compiling, which prevents us from referencing any system headers entirely (even ones provided by the compiler like `stdint.h` are redirected to the PAL). This problem is not solved as a part of this PR and will need to be addressed in the future going forward - I think this could mean breaking the GC's dependency on the PAL entirely before this interface can be fully implemented.

It's also important to note that none of these changes are building by default and only are in effect when `FEATURE_STANDALONE_GC` is defined. In addition, `FEATURE_STANDALONE_GC` will not actually compile on Unix yet due to an `#error` directive on the implementation.

Finally, this PR contains some additional changes (in `ceemain.cpp`, `gcenv.ee.cpp`, `gcenv.ee.h`, and `gcenv.h`) that were added in the PR that introduced the standalone `GCToEEInterface` but were lost in a merge conflict. These are in a separate commit.

### CODE REVIEW

**Comment by swgillespie** [comment]
cc @sergiy-k @jkotas @Maoni0 @janvorli - PTAL if interested? The implementation isn't at 100% parity on Windows yet but I figured this was a good checkpoint to merge without the diff getting really large.

File: src/gc/gcenv.windows.cpp [hunk 1]
```diff
@@ -0,0 +1,697 @@
...
+
+// Reserve virtual memory range.
+// Parameters:
+//  address   - starting virtual address, it can be NULL to let the function choose the starting address
+//  size      - size of the virtual memory range
+//  alignment - requested memory alignment, 0 means no specific alignment requested
+//  flags     - flags to control special settings like write watching
+// Return:
+//  Starting virtual address of the reserved range
+void* GCToOSInterface::VirtualReserve(void* address, size_t size, size_t alignment, uint32_t flags)
```
Comment (janvorli) [inline src/gc/gcenv.windows.cpp]: It seems we should remove the address parameter - we don't use it here and GC always passes in 0. Seems like I have overlooked it while creating the GCToOSInterface in the past.

File: src/gc/gcenv.windows.cpp [hunk 2]
```diff
@@ -0,0 +1,697 @@
...
+{
+    // TODO(segilles) platform detection
+    return false;
+}
+
+// For coresys we need to look for an API in some apiset dll on win8 if we can't find it  
+// in the traditional dll.
+HINSTANCE LoadDllForAPI(const WCHAR* dllTraditional, const WCHAR* dllApiSet)
+{
+    HINSTANCE hinst = LoadLibraryEx(dllTraditional, nullptr, 0);
```
Comment (jkotas) [inline src/gc/gcenv.windows.cpp]: Pass `LOAD_LIBRARY_SEARCH_SYSTEM32` as flags - to make sure that the system is not looking for the .dlls in random places.

File: src/gc/gcenv.windows.cpp (hunk 1 above)
Comment (swgillespie) [inline src/gc/gcenv.windows.cpp]: will do!

File: src/gc/gcenv.windows.cpp (hunk 2 above)
Comment (swgillespie) [inline src/gc/gcenv.windows.cpp]: will do!

File: src/gc/gcenv.windows.cpp [hunk 3]
```diff
@@ -0,0 +1,697 @@
...
+{
+    LIMITED_METHOD_CONTRACT;
+
+    // The limit was cached already
+    if (g_RestrictedPhysicalMemoryLimit != (size_t)UINTPTR_MAX)
+        return g_RestrictedPhysicalMemoryLimit;
+
+    size_t job_physical_memory_limit = (size_t)UINTPTR_MAX;
+    BOOL in_job_p = FALSE;
+#ifdef FEATURE_CORECLR
```
Comment (jkotas) [inline src/gc/gcenv.windows.cpp]: This should not be under `FEATURE_CORECLR` ifdef - the CoreCLR path should be used all the time.

Comment (swgillespie) [inline src/gc/gcenv.windows.cpp]: ok!

Comment (swgillespie) [inline src/gc/gcenv.windows.cpp]: are you referring to the all of the DLL loading behavior in this function or this particular ifdef?

Comment (jkotas) [inline src/gc/gcenv.windows.cpp]: All `FEATURE_CORECLR` ifdefs in this file.

**Comment by Maoni0** [comment]
LGTM. thanks Sean!

**Comment by janvorli** [comment]
LGTM

**Comment by sergiy-k** [comment]
LGTM

**Comment by swgillespie** [comment]
thanks!

**Pull request merged**